src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs

```diff
@@ -805,7 +815,7 @@ internal JsonTypeInfo JsonTypeInfo
         internal JsonNumberHandling? EffectiveNumberHandling { get; set; }
 
         //  Whether the property type can be null.
-        internal bool PropertyTypeCanBeNull { get; }
+        internal abstract bool PropertyTypeCanBeNull { get; }
```

Comment:
it's used in `ReadJsonAndSetMember`, IMO good enough reason to cache this

Reply:
I think it would make sense to run perf tests for this change to make sure that's the case. I'm not that familiar with what JIT exactly does but gut tells me that using virtual properties rather than non-virtual should be faster

Reply:
I'm seeing the Bar is in fact returning const but I don't see how method call is faster than reading field/property. Should we at least inline it?

Reply:
[actually fine ](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEAygBYQBOADgDITAB0ASgK4wJRgCmAbiKkymXB268BrAMIB7MPSgAbfrQrqAblADG/FMMIiAzGRYoEtCLoRjsZAGLz5RAN5EyXsWYtWbdsAuKmQAQnRkbgDm/AiCAL5EicaEANoAUlAIAOL8MOp6ABQIAJ70/PIAZoWwCACUdXBkmTl5BbrFZRXVtQD8DQC6pmQo/BBqACb2Ti4APAAqAHxkIDOuhB6E3j5k8lrqtFAT/GRB8iHhtGQAvMvHlRCcKgiF83VkUChkME8qRtuebyYMxnC50BzXMj3R7PV7vT7fX7/byArzA07BMjzAwvd63MJ0ZFokGY7GWbCFPHLS7YIzJMhAA=)

```
Foo`1[[System.Nullable`1[[System.Int32, System.Private.CoreLib]], System.Private.CoreLib]].Test()
    L0000: cmp [ecx], ecx
    L0002: call Foo`1[[System.Nullable`1[[System.Int32, System.Private.CoreLib]], System.Private.CoreLib]].get_Bar()
    L0007: ret
```

vs

```
Foo`1[[System.Nullable`1[[System.Int32, System.Private.CoreLib]], System.Private.CoreLib]].Test2()
    L0000: movzx eax, byte ptr [ecx+4]
    L0004: ret
```

Reply:
I think best option would be to just use it directly everywhere rather than any variables

Reply:
[even better solution](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEAygBYQBOADgDITAB0ASgK4wJRgCmAbiKkymXB268BrAMIB7MPSgAbfrQrqAblADG/FMMIiAzGRYoEtCLoRjsZAGLz5RAN5EyXsWYtWbdsAuKmQAQnRkbgDm/AiCAL5EicaEANoAUlAIAOL8MOp6ABQIAJ70/PIAZoWwCACUdXBkmTl5BbrFZRXVtQD8DQC6pmQo/BBqACb2Ti4APAAqAHxkIDOuhB6E3j5k8lrqtFAT/GRB8iHhtGQAvMvHlRCcKgiF83VkUChkME8qRtuebyYMxnC50BzXMj3R7PV7vT7fX7/byArypACysSo8gmAEklCpCpiENi8QSAPL0XjyGAoVgAQSiUVoBhQUH2uJgKlgsCidSGWyBIOCYToJkK71uUP4DyeLzeHy+PxUf1RZDVwNOIvmBhekuWl2RXk1oLIOss2AlNwN4KNO1N5oQ4v1otozrtGuF5zNuoA+hMoCzbFapdC5XDFYiVUZkmQgA=)

See Test3, so basically we should change that to a method and add aggressive in-lining - at this point it can probably be abstract as well since it will be devirtualized

Reply:
output of Test3:

```
Foo`1[[System.Nullable`1[[System.Int32, System.Private.CoreLib]], System.Private.CoreLib]].Test3()
    L0000: mov eax, 1
    L0005: ret
```

so basically treated as const after devirtualization. For some reason not happening for properties (also I cannot do aggressive inlining for properties)